### ISSUE TITLE
Semi-supervised learning with YOLOv8

### ISSUE THREAD

**Issue by AdriCon02** [ultralytics/ultralytics#10248]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi, 

my question may be a bit long and tricky, but I will try to be as clear as possible, besides my knowledge of ML is not very extensive.

Apart from being quite unbalanced in terms of classes, I am facing a dense object detection dataset that is not fully labeled, and as YOLOv8 introduces the background class, my results of mAP, precision, recall and confusion matrix doesn't seem to be very representative, providing a very high false positive rate (when the model may be working very well, just not all elements are labeled).

Another problem is that this dataset is related to the medical field, so I cannot complement the labels since I do not have the necessary knowledge.

My doubt is in how I can face this kind of problems with YOLOv8, either from the following ideas that I have come up with:

-Can semi-supervised learning strategies be carried out with YOLOv8? 

-If so, how can I then have an accuracy metric of my model if I am going to work with a batch of data that a priori has not been labeled by a subject matter expert? (This is more related to a concept of metrics accuracy, not with YOLOv8, sorry...).

-Can I make an accuracy metric with my current YOLOv8 model that only takes into account predictions about what HAS been labeled, so I would have a real, undistorted metric?

Finally, I have also thought about applying active learning strategies, but as I mentioned before, I don't have the necessary knowledge to intervene in a Human-In-The-Loop system. I hope you have understood my question and I am sorry if it is tedious to read, but it is driving me a bit crazy! Thank you.

Adrián

**Comment by glenn-jocher** [comment]
@AdriCon02 hi Adrián 👋,

Your question is very insightful, and it's great to see your proactive approach towards handling a semi-supervised learning scenario with YOLOv8, especially in a critical field like medical object detection.

- **Semi-supervised Learning with YOLOv8**: Currently, YOLOv8 does not natively support semi-supervised learning out of the box. However, this doesn't mean you're out of options. You can use labeled data for supervised learning and explore data augmentation or synthetic data generation techniques to enrich your dataset.

- **Metrics for Semi-supervised Learning**: Evaluating the accuracy of a model trained on partially labeled data is indeed challenging. One approach could be to focus on metrics computed only on the labeled subset of your data. This means evaluating your model's performance on a test set that is fully labeled, ensuring your metrics reflect the model's predictions on known ground truth.

  Here's a pseudo example for filtering predictions for evaluation:
  ```python
  # Assuming `predictions` are your model's outputs and `labels` are your ground truth
  filtered_predictions = filter_out_unlabeled(predictions, labels)
  evaluate_model(filtered_predictions, labels)
  ```
  Remember, this approach assesses performance solely on the labeled data, which may not fully represent your model's potential accuracy across all data.

- **Active Learning**: Given the constraints with medical datasets, it's understandable that intervening directly may not be an option. Instead, consider collaborating with medical professionals who can validate and label the most uncertain predictions made by your model. This targeted approach to increasing your labeled dataset can significantly improve model performance over time.

Your initiative to explore these methodologies is commendable, and while YOLOv8 might have limitations in directly supporting semi-supervised learning, leveraging external strategies to enrich your dataset and refine the evaluation metrics can pave a way forward.

If experimenting with these approaches or integrating additional tools feels overwhelming, seeking collaborations with domain experts for dataset enrichment and validation could be a worthwhile pathway.

Best of luck with your project, Adrián! Your work is incredibly important, and navigating these challenges is a big step towards impactful solutions. 🚀

**Comment by AdriCon02** [comment]
Hi Glenn!

Thanks for your quick response, you gave me some valuable tips to keep working and trying to improve that mAP! 

Just one question more, regarding the pseudocode you provided, it is quite difficult to make an approach for filtering the non-labeled data correspondant predictions, so I came up with the following idea, but I´m not sure if it would work...

What I thought about is to compute the IoU value of each predicition with its nearest labeled data, then filter the annotations by applying a personalized threshold (for example, assume that predictions with an IoU lower than 0.2 mean that it is not matched to any labeled data), then I'd only have the labeled data predictions. 

Let me know what do you think about this. Thanks in advance.

Adrián

**Comment by glenn-jocher** [comment]
@AdriCon02 hey Adrián! 😊

I'm glad to hear the tips were helpful! Your idea about using IoU to filter out predictions related to non-labeled data is actually a pretty solid approach. By computing the IoU between each prediction and its nearest ground truth and applying a custom threshold, you're essentially focusing on evaluating only those predictions that have a meaningful overlap with labeled data. Your pseudocode idea aligns well with best practices for dealing with semi-supervised scenarios.

Here’s a brief example snippet to illustrate your idea:
```python
iou_threshold = 0.2
filtered_predictions = [pred for pred in predictions if compute_iou(pred, nearest_label(pred)) > iou_threshold]
```
Make sure `compute_iou()` calculates the IoU, and `nearest_label(pred)` finds the closest labeled ground truth to a given prediction.

Remember, fine-tune the threshold based on your specific dataset characteristics and task requirements.

Keep up the innovative thinking! 🌟

**Comment by AdriCon02** [comment]
Hi Glenn!

Finally, I implemented the pseudocode you provided. I now have only the predictions that are supposed to correspond to the original annotated elements in the image. The main issue now that I'm facing is to implement the original validation process to obtain mAP.

Is there any way to use the yolo=detect task=val command but manually introducing these new predictions? 

If you or anyone comes up with any new idea, I'd be extremely grateful!

Greetings,
Adrián

**Comment by glenn-jocher** [comment]
Hey Adrián! 👋

Great work on implementing the filtering process! For running the validation with your filtered predictions, the validation in YOLOv8 typically expects model outputs directly rather than post-processed predictions.

However, you can manually evaluate the filtered predictions by comparing them against your test set annotations, calculating the mAP yourself. You might need to write a custom script for this that computes IoU between your predictions and the ground truths and then calculates the precision-recall curve.

A simple version might look something like this (ensure to install the required libraries):

```python
from sklearn.metrics import average_precision_score

# Assume `y_true` to be ground truth labels, `y_scores` to be prediction scores
mAP = average_precision_score(y_true, y_scores)
print(f"mAP: {mAP}")
```

Remember, while this is a simplified example, you might need a more detailed approach depending on your exact requirements.

Keep pushing the boundaries, Adrián! 🚀

**Comment by AdriCon02** [comment]
Hi Glenn!

I think it will be easier if I show you an example of what I have:

**Annotations: bbox & classes**

```
tensor([[385, 145, 475, 322],
        [450, 600, 520, 640],
        [345, 555, 400, 630],
        [320, 480, 380, 552],
        [510, 475, 575, 582],
        [500, 355, 580, 470],
        [495, 260, 565, 327],
        [290, 240, 370, 325],
        [535, 160, 600, 250],
        [455,  85, 530, 187],
        [360,  85, 415, 175],
        [295, 150, 365, 225],
        [230, 195, 300, 285],
        [150,  70, 205, 160],
        [105,   0, 160,  65],
        [280,   0, 330,  85],
        [365,   0, 416,  35]])
tensor([1, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5])
```

**Predictions: pred_bbox & pred_classes**

```
tensor([[2.9084e+02, 1.5278e+02, 3.6680e+02, 2.2489e+02],
        [2.3032e+02, 1.9479e+02, 3.0133e+02, 2.8384e+02],
        [2.7754e+02, 0.0000e+00, 3.3222e+02, 8.3485e+01],
        [2.8931e+02, 2.3845e+02, 3.6875e+02, 3.2518e+02],
        [4.5362e+02, 8.2276e+01, 5.3044e+02, 1.9127e+02],
        [4.9984e+02, 3.5417e+02, 5.8005e+02, 4.7040e+02],
        [4.9572e+02, 2.5920e+02, 5.6327e+02, 3.3306e+02],
        [3.1946e+02, 4.8280e+02, 3.8158e+02, 5.5538e+02],
        [3.4518e+02, 5.5278e+02, 4.0229e+02, 6.3138e+02],
        [5.1165e+02, 4.7789e+02, 5.7822e+02, 5.8340e+02],
        [3.6088e+02, 8.1780e+01, 4.1354e+02, 1.7669e+02],
        [3.5213e+02, 3.1078e+02, 4.0942e+02, 3.7842e+02],
        [3.8572e+02, 1.4554e+02, 4.7599e+02, 3.3266e+02],
        [5.3264e+02, 1.6080e+02, 5.9805e+02, 2.5010e+02],
        [3.6576e+02, 1.6159e-02, 4.1678e+02, 3.2970e+01],
        [4.4803e+02, 6.0072e+02, 5.2097e+02, 6.3959e+02],
        [1.0445e+02, 6.3417e-01, 1.6098e+02, 6.2065e+01],
        [3.2167e+02, 4.0025e+02, 3.9641e+02, 4.7345e+02]])
tensor([5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 1., 5., 5., 5., 5., 5.])
```

**Filtered Predictions: filtered_pred_bbox & filtered_pred_classes**

```
tensor([[2.9084e+02, 1.5278e+02, 3.6680e+02, 2.2489e+02],
        [2.3032e+02, 1.9479e+02, 3.0133e+02, 2.8384e+02],
        [2.7754e+02, 0.0000e+00, 3.3222e+02, 8.3485e+01],
        [2.8931e+02, 2.3845e+02, 3.6875e+02, 3.2518e+02],
        [4.5362e+02, 8.2276e+01, 5.3044e+02, 1.9127e+02],
        [4.9984e+02, 3.5417e+02, 5.8005e+02, 4.7040e+02],
        [4.9572e+02, 2.5920e+02, 5.6327e+02, 3.3306e+02],
        [3.1946e+02, 4.8280e+02, 3.8158e+02, 5.5538e+02],
        [3.4518e+02, 5.5278e+02, 4.0229e+02, 6.3138e+02],
        [5.1165e+02, 4.7789e+02, 5.7822e+02, 5.8340e+02],
        [3.6088e+02, 8.1780e+01, 4.1354e+02, 1.7669e+02],
        [3.8572e+02, 1.4554e+02, 4.7599e+02, 3.3266e+02],
        [5.3264e+02, 1.6080e+02, 5.9805e+02, 2.5010e+02],
        [3.6576e+02, 1.6159e-02, 4.1678e+02, 3.2970e+01],
        [4.4803e+02, 6.0072e+02, 5.2097e+02, 6.3959e+02],
        [1.0445e+02, 6.3417e-01, 1.6098e+02, 6.2065e+01]])
tensor([5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 1., 5., 5., 5., 5.])
```
Please notice that the filtered prediction is shorter in size than the original one, consequence of the filtering process. The code you gave me needs probabilities as an input, which is something I do not have. Any new tip for calculating mAP with these metadata? Thanks in advance.

Adrián

**Comment by AdriCon02** [comment]
This is just an example of 1 image sorry, I would like to have the mAP for the whole validation dataset :)

**Comment by glenn-jocher** [comment]
Hello Adrián! 👋

To calculate mAP across your entire validation dataset, you'll want to aggregate the predictions and corresponding ground truth for all images in the dataset, and then compute the mAP for these aggregated results. Here’s how you can approach it:

1. Loop through all images in your validation dataset.
2. Apply your prediction and filtering process to each image.
3. Collect all filtered predictions and their corresponding ground truths.
4. Calculate mAP using these collected predictions and ground truths.

Here's a very simplified pseudocode snippet to guide you:

```python
all_predictions = []
all_ground_truths = []

for image in validation_dataset:
    predictions = model.predict(image)
    filtered_preds = filter_predictions(predictions)
    all_predictions.append(filtered_preds)
    all_ground_truths.append(ground_truths_for_image)

map_score = calculate_map(all_predictions, all_ground_truths)
print(f"mAP for the entire validation dataset: {map_score}")
```
Remember to replace functions like `model.predict`, `filter_predictions`, and `calculate_map` with your actual implemented functions.

Let me know if you need further guidance! 😊
